src/storage/src/render/upsert.rs

```diff
@@ -175,17 +202,32 @@ where
         consolidation::consolidate(&mut snapshot);
 
         // The main key->value used to store previous values.
-        let mut state = HashMap::new();
+        let mut state = StatsState::new(InMemoryHashMap::default(), upsert_shared_metrics);
 
-        // A re-usable buffer of changes, per key. This is a `std`
-        // `HashMap` so we can use `HashMap::drain` below.
-        #[allow(clippy::disallowed_types)]
-        let mut commands_state = std::collections::HashMap::new();
+        // A re-usable buffer of changes, per key. This is
+        // an `IndexMap` because it has to be `drain`-able
+        // and have a consistent iteration order.
+        let mut commands_state = indexmap::IndexMap::new();
+        let mut multi_get_scratch = Vec::new();
 
-        for ((key, value), diff) in snapshot {
+        // Rehydrate the upsert state (and bump some stats), even if the snapshot is empty.
+        let snapshot = snapshot.into_iter().map(|((key, value), diff)| {
             assert_eq!(diff, 1, "invalid upsert state");
-            state.insert(key, value);
-        }
+            (key, Some(value))
+        });
+
+        let now = Instant::now();
+        let snapshot_size = snapshot.len();
+        state
+            .multi_put(snapshot)
+            .await
+            .expect("hashmap impl to not error");
```

Comment:
I would suggest making the trait never return an error since there is no HashMap/RocksDb error that can be handled so it is only adding `.unwrap()` noise to the caller.

Reply:
We don't want to unwrap in the implementation, we want to report the error to the user and restart the dataflow, so the error does need to be reported back!